### PR TITLE
Mo/issue 208

### PR DESCRIPTION
**Pull request by MarcoOl94**
Introduced new lock to avoid deadlock problem and related GUnit test.

### CODE REVIEW

**Review by abi87 — APPROVED** [review]
I contributed to the task, hence all of my comments are embodied in code.
Results of investigations are in Monday task OA64.

The unit test really helps isolating the root cause of deadlock, allowing clear investigation with further tools (we used helgrind).  The test completes all right when no deadlock appears, it hangs upon deadlock; I think it is worth keeping it there as embodiment of acquired knowledge.

**Comment by cronicc** [comment]
Some upstream issues related to deadlocks https://github.com/zcash/zcash/issues/3825, https://github.com/zcash/zcash/issues/3823 and a recently merged PR https://github.com/zcash/zcash/pull/3932.

`CONFIGURE_FLAGS='--enable-debug' zcutil/build.sh` seems to be of particular interest for debugging purposes.

**Comment by abi87** [comment]
> 
> 
> Some upstream issues related to deadlocks [zcash/zcash#3825](https://github.com/zcash/zcash/issues/3825), [zcash/zcash#3823](https://github.com/zcash/zcash/issues/3823) and a recently merged PR [zcash/zcash#3932](https://github.com/zcash/zcash/pull/3932).
> 
> `CONFIGURE_FLAGS='--enable-debug' zcutil/build.sh` seems to be of particular interest for debugging purposes.

Cool feature I knew nothing about, thanks @cronicc . 
We went for [helgrind](https://valgrind.org/docs/manual/hg-manual.html) checks with unit test written by @MarcoOl94 to isolate relevant operation. 
Next time we'll start by CONFIGURE_FLAGS='--enable-debug'

File: src/wallet/gtest/test_deadlock.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    boost::filesystem::path pathTemp = boost::filesystem::temp_directory_path()
+            / boost::filesystem::unique_path();
+    boost::filesystem::create_directories(pathTemp);
+    mapArgs["-datadir"] = pathTemp.string();
+
+    //create wallet
+    SelectParams(CBaseChainParams::TESTNET);
+    bool fFirstRun = true;
+    CWallet walletMain("deadlock_ut_wallet.dat");
+    DBErrors nLoadWalletRet = walletMain.LoadWallet(fFirstRun);
```
Comment (alsala) [inline src/wallet/gtest/test_deadlock.cpp]: nLoadWalletRet is not used

File: src/wallet/gtest/test_deadlock.cpp [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    auto sk = libzcash::SpendingKey::random();
+    walletMain.AddSpendingKey(sk);
+
+    //create block
+    CBlock block;
```
Comment (alsala) [inline src/wallet/gtest/test_deadlock.cpp]: Maybe the obj can be created in the function thread (instead of passing in the thread creation)

**Review by albertog78 — APPROVED** [review]

**Review by cronicc — APPROVED** [review]

**Pull request merged**